### PR TITLE
ext_proc: Implement STREAMED body processing mode

### PR DESCRIPTION
**Pull request by gbrail**
Commit Message: Implement STREAMED body processing mode in ext_proc

Additional Description: Implement this mode, which allows an external processor to receive the body one chunk at a time and examine or modify the body while it is being processed.

Risk Level: Medium. The mode is only engaged if used. However an external processor must be coded carefully, particularly if it expects to stream both request and response bodies simultaneously, since each will proceed in its own ordering.

Testing: New unit and integration tests.

Docs Changes: Updated proto docs to reflect that the mode is now supported.

Release Notes: External processing servers may now use the STREAMED processing mode. In this mode, chunks of the body are forwarded to the external processing server when they arrive. Depending on how the upstream system is implemented, request body chunks may be delivered before or after the response headers, and request and response body chunks may be interleaved if the upstream system delivers them that way. An external processor should be carefully coded so that it does not assume that a particular ordering will be implemented.

### CODE REVIEW

**Comment by gbrail** [comment]
@ikepolinsky

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -331,24 +366,31 @@ void Filter::onReceiveMessage(std::unique_ptr<ProcessingResponse>&& r) {
...
   case ProcessingResponse::ResponseCase::kRequestBody:
+    ENVOY_LOG(debug, "Received RequestBody response");
     message_handled = decoding_state_.handleBodyResponse(response->request_body());
     break;
   case ProcessingResponse::ResponseCase::kResponseBody:
+    ENVOY_LOG(debug, "Received ResponseBody response");
     message_handled = encoding_state_.handleBodyResponse(response->response_body());
     break;
   case ProcessingResponse::ResponseCase::kRequestTrailers:
+    ENVOY_LOG(debug, "Received RequestTrailers response");
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:385]: Nit: could have a enum-to-string converter to make this less verbose.

File: source/extensions/filters/http/ext_proc/processor_state.h [hunk 2]
```diff
@@ -26,6 +37,8 @@ class ProcessorState : public Logger::Loggable<Logger::Id::filter> {
     HeadersCallback,
     // Waiting for a "body" response in buffered mode
     BufferedBodyCallback,
+    // Waiting for a streamed chunk
+    StreamedBodyCallback,
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.h:41]: For my education, why can't we get into the state where we're waiting for both headers and streamed chunk, e.g. at the start of a stream when we received headers/data independently as decoder callbacks?

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 3]
```diff
@@ -171,11 +172,43 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
...
+      break;
+    }
+
+    auto next_chunk = std::make_unique<QueuedChunk>();
+    // Clear the current chunk and save it on the queue while it's processed
+    next_chunk->data.move(data);
+    next_chunk->end_stream = end_stream;
+    // Send the chunk, and ensure that we have watermarked so that we don't overflow
+    // memory while waiting for responses.
+    state.requestWatermark();
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:192]: We set watermark on every single chunk?

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc [hunk 4]
```diff
@@ -261,6 +407,286 @@ TEST_P(StreamingIntegrationTest, GetAndProcessBufferedResponseBody) {
...
+
+// Send a large HTTP POST, and expect back an equally large reply. Stream both and verify
+// that we got back what we expected.
+TEST_P(StreamingIntegrationTest, PostAndProcessStreamBothBodies) {
+  const uint32_t send_chunks = 10;
+  const uint32_t chunk_size = 11000;
+  uint32_t request_size = send_chunks * chunk_size;
+  uint32_t response_size = 1700000;
+
+  test_processor_.start(
```
Comment (htuch) [inline test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: These gRPC servers are kind of large, hard to read and their logic is not inline with the rest of the test. Is there a reason we went this path vs. what is done in the rest of the e2e integation tests in Envoy? Can more boiler plate be factored out? Can they be made more declarative?

**Review by htuch — CHANGES_REQUESTED** [review]
Thanks, this is super cool.
/wait

**Comment by htuch** [comment]
@ikepolinsky would appreciate your take on this PR; since you will be writing the fuzzing against this, it's probably a good idea to make sure it's testable and matches your understanding.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:192]: That call is idempotent (that it, it has a flag so that it only calls the corresponding method on the stream context if it is not already set.) We leave the watermark "high" as long as there are stream messages outstanding.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 4 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: "streaming_integration_test" is a new test I introduced in the last PR. Those other tests are still there, in "ext_proc_integration_test." 

These tests use a real gRPC server rather than simulating it through the integration framework, and it runs in a separate thread. To test streaming I wanted to be able to stream chunks of the body in both directions and the integration framework seems more designed to do things in a linear way -- send body from downstream, read it upstream, send it back upstream, read the result -- and in reality all those things may happen at once.

I personally find this "grpc server" style much easier for me to read and write because the ext_proc server that we're testing is the same server that a user would actually create, rather than a simulation, and IMO it's closer to replicating the behavior of the real thing in which chunks of data are passing through the filter, to the server, and back again, all in parallel.

Comment (gbrail) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: But yes, they could be shrunk with some helper functions.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:385]: Is there a canonical Envoy way to do this, such as an "operator <<" or a ToString method? I didn't find anything in filter.h for the filter callback enums, for instance.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:192]: I'm hoping that resolves this but let me know otherwise.

Comment (htuch) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:192]: This seems a bit excessive. Shouldn't we set the high watermark when we trip over an actual watermark based on number of outbound requests?

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 4 above)
Comment (htuch) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: I think at least reducing boiler plate would be good. It's hard to follow the test really as it's currently written.

File: source/extensions/filters/http/ext_proc/processor_state.h (hunk 2 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.h:41]: We can be in that state -- in that case, we are in "HeadersCallback" state and we buffer the data until the headers callback returns, then there is logic in ProcessorState::handleHeadersResponse to continue processing the buffered body depending on the body mode that's selected -- which might not have been known until the headers response came back.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:192]: We can keep track of how much data is outstanding and only raise the watermark when it's above a threshold. I'm not sure what the threshold should be -- should it be the buffer limit that the user configured, or should it be half that? This buffering is happening entirely within this filter, so if there's other buffering happening elsewhere in Envoy I don't want to make the problem too much worse.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 4 above)
Comment (ikepolinsky) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: Correct me if I'm wrong. My understanding is the gRPC server is necessary because it introduces different interleaving of headers, body chunks, etc. and the functions provided in the HTTPIntegration class are "all or nothing" sending a full request and waiting for a full response. The downside of this approach is that we cannot recreate a specific interleaving if it causes an issue.

If the above is true, this is extremely related to my fuzzing goal. My goal is to be able to control how many chunks are sent, when they are sent, when the mock gRPC server responds to the filter, the number of chunks the upstream responds with, and when the upstream responds. Although, I'm not sure if the filter is deterministic when it has both data from the downstream and upstream waiting to be processed.

Is this level of granularity something that would be useful for generic ext_proc integration tests, @gbrail? One downside that immediately comes to mind is that writing tests where you manually specific the downstream, upstream, and gRPC server actions will be very tedious and possibly result in long test cases.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (htuch) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:192]: Yeah, there are two things here:
1. When we raise the watermark, we're potentially causing HTTP flow control event logic to be triggered. 
2. There should be some high/low watermark distinction to avoid oscillating on (1).

I think if we track the number/size of outgoing buffered resources and provide some basic control over the upper bound we can consult it here, doing something like low watermark = high watermark / 2.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 4 above)
Comment (htuch) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:485]: I think that's a fair description; you lose thread interleaving determinism unless the RPC sequences effectively serialize this, which I don't think is true with the `ext_proc` async RPC design in general.

File: source/extensions/filters/http/ext_proc/processor_state.h (hunk 2 above)
Comment (htuch) [reply source/extensions/filters/http/ext_proc/processor_state.h:41]: Ack; can you add some comments here explaining ^^ and this can be resolved..

**Comment by gbrail** [comment]
Thanks for your comments -- I have some improvements I'm working on but I have one test failure to address before they're ready.

File: source/extensions/filters/http/ext_proc/processor_state.h (hunk 2 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.h:47]: The comments are better now -- hope that helps.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 3 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: I changed it to only raise the watermark when the total size of the queue is greater than the buffer limit / 2 -- that reduces the amount of watermarking a lot.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 4 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:471]: I made the tests a bit better with a helper function and by switching to standard GTest macros. For this PR I added both these "streaming integration tests" and also new mock-based tests with a deterministic order in filter_test.cc, so I think that we're fairly well covered, but I'm also looking forward to the fuzz tests.

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 5]
```diff
@@ -160,22 +180,52 @@ FilterDataStatus Filter::onData(ProcessorState& state, Buffer::Instance& data, b
...
+      // Fall through
+      break;
+    }
+
+    // Send the chunk on the gRPC stream
+    sendBodyChunk(state, data, ProcessorState::CallbackState::StreamedBodyCallback, end_stream);
+    // Move the data to the queue and optionally raise the watermark.
+    state.enqueueStreamingChunk(data, end_stream, true);
+
+    // At this point we will continue, but with no data, because that will come later
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:213]: Why do you continue with no data?

File: source/extensions/filters/http/ext_proc/processor_state.cc [hunk 6]
```diff
@@ -67,44 +68,116 @@ bool ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
...
+          filter_.sendBufferedData(*this, ProcessorState::CallbackState::BufferedBodyCallback,
+                                   true);
         }
 
         // Otherwise, we're not ready to continue processing because then
         // we won't be able to modify the headers any more, so do nothing and
         // let the doData callback handle body chunks until the end is reached.
+        clearWatermark();
+        return true;
+
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:80]: Nit: gratuitous blank line

File: source/extensions/filters/http/ext_proc/processor_state.cc [hunk 7]
```diff
@@ -128,18 +204,50 @@ bool ProcessorState::handleTrailersResponse(const TrailersResponse& response) {
...
 
+void ProcessorState::enqueueStreamingChunk(Buffer::Instance& data, bool end_stream,
+                                           bool delivered) {
+  bytes_enqueued_ += data.length();
+  auto next_chunk = std::make_unique<QueuedChunk>();
+  next_chunk->data.move(data);
+  next_chunk->end_stream = end_stream;
+  next_chunk->delivered = delivered;
+  chunks_for_processing_.push_back(std::move(next_chunk));
+  if (bytes_enqueued_ > bufferLimit() / 2) {
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:221]: Is this the opposite of what is wanted? We should set watermark when we go over the high watermark and clear when we drop below the low, with low watermark = high watermark / 2.

File: source/extensions/filters/http/ext_proc/processor_state.cc [hunk 8]
```diff
@@ -67,44 +68,116 @@ bool ProcessorState::handleHeadersResponse(const HeadersResponse& response) {
...
+      });
+      clearWatermark();
+      callback_state_ = CallbackState::Idle;
+      should_continue = true;
+
+    } else if (callback_state_ == CallbackState::StreamedBodyCallback ||
+               callback_state_ == CallbackState::StreamedBodyCallbackFinishing) {
+      bool delivered_one = false;
+      while (auto queued_chunk = dequeueStreamingChunk(delivered_one)) {
+        // Loop through queue in case some of it is chunks that were never
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.cc:151]: It's a little unclear to me when streaming can degenerate into general buffering. I think above, before headers are received back from the ext_proc server, this can happen. Here we dequeue chunks, but this is subject to just the stream buffer that we are watermarking?

File: source/extensions/filters/http/ext_proc/processor_state.h [hunk 9]
```diff
@@ -16,29 +20,50 @@ namespace ExternalProcessing {
...
+    // to receive data frames for a message.
     HeadersCallback,
-    // Waiting for a "body" response in buffered mode
+    // Waiting for a "body" response in buffered mode.
     BufferedBodyCallback,
-    // and waiting for a "trailers" response
+    // Waiting for a "body" response in streaming mode.
+    StreamedBodyCallback,
+    // Waiting for a "body" response in streaming mode in the special case
+    // in which the processing mode was changed while there were outstanding
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/processor_state.h:49]: When would we sensible change the processing mode during streaming?

**Review by htuch — CHANGES_REQUESTED** [review]
Sorry about the delay, holiday week.
/wait

File: source/extensions/filters/http/ext_proc/processor_state.h (hunk 9 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.h:49]: Sensibly, I could see a case in which a filter might want to inspect only the beginning of a large message body. As another example, I was thinking of a caching solution recently in which we'd want to cache message bodies, but only if the message body is smaller than a certain size. Once a processor sees that the body will be too large, it might want to change the processing mode to streamline things if it no longer cares.

The other reason is because we allow that in the protocol now. If we decide instead to ignore processing mode changes while streaming a body then we likely need to add some more logic elsewhere.

Comment (GuyLewin) [reply source/extensions/filters/http/ext_proc/processor_state.h:49]: @gbrail it could be especially useful if you're looking for a specific substring in a request / response body, and once you found it you can ignore the rest (since a policy has been already activated)

File: source/extensions/filters/http/ext_proc/processor_state.cc (hunk 8 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc:151]: Streaming degenerates into buffering if, while we were waiting for the response to the "headers" message from the external processor, the entire message body comes in to the filter and is buffered. At that point, whether the processor requests "streamed" or "buffered" mode it's the same thing since we already have the whole message in the buffer.

File: source/extensions/filters/http/ext_proc/processor_state.cc (hunk 7 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc:221]: I'm checking in a change to the watermarking logic in the streaming case so it will raise the watermark when it goes over the buffer limit, and lower it when it goes below half the buffer limit. I was confused because I'm not sure if the buffering that this filter is doing is in addition to whatever else the rest of Envoy does. Is there a chance that by not requesting watermarking until "bufferSize" messages are in the queue inside this filter, we are actually doubling the amount of memory used since presumably the rest of Envoy is buffering too?

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 5 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:213]: We had some conversations about that in the issue that I opened before coding this up. I can find the original discussion, but without making big changes to the way that the filter chain is processed for all filters, there's no way for this filter to tell Envoy to stop sending new chunks of data right away, which in turn makes it hard to let an external processor modify each chunk individually. And even if we could do that, then it would introduce a lot of latency because we'd be pausing the filter flow after each chunk to wait for the processor.

This way, there is a flow of data to the external processor, and a flow of data back. We do this by moving the data out of each chunk and letting it keep flowing through the filter chain, and then introducing new chunks into the filter chain when they come back from the processor.

Comment (htuch) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:213]: This is complicated enough that I think linking to the thread in the comment would make sense; it's counter-intuitive that we would `Continue` (meaning pass to the next filter in the pipeline) without a chunk.

File: source/extensions/filters/http/ext_proc/processor_state.cc (hunk 7 above)
Comment (htuch) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: Envoy has multiple buffers subject to watermarks and in fact each filter may do its own buffering. To know the actual memory in use you need to look at a specific filter chain and multiple watermark limits by the number of buffers on the path.

It would be consistent for the watermark/buffer limits to remain constant, we should be referencing high/low watermark limits.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc [hunk 10]
```diff
@@ -235,30 +386,248 @@ TEST_P(StreamingIntegrationTest, GetAndProcessBufferedResponseBody) {
...
+          ProcessingResponse response;
+          ASSERT_TRUE(stream->Read(&message));
+          if (message.has_response_headers()) {
+            // Expect only one response headers message, with a good status.
+            EXPECT_FALSE(saw_response_headers);
+            EXPECT_THAT(message.response_headers().headers(),
+                        SingleProtoHeaderValueIs(":status", "200"));
+            saw_response_headers = true;
+            response.mutable_response_headers();
+
```
Comment (htuch) [inline test/extensions/filters/http/ext_proc/streaming_integration_test.cc:500]: Nit: quite a few places in this PR have this gratuitous blank line at the end of an `if` block. Prefer compact style as per rest of code base.

**Review by htuch — COMMENTED** [review]
Thanks, this reads well to me now, the only two things before shipping are nits and maybe getting some better docs on the thread around HTTP filter chain resumption while chunks are out for processing.

File: test/extensions/filters/http/ext_proc/streaming_integration_test.cc (hunk 10 above)
Comment (gbrail) [reply test/extensions/filters/http/ext_proc/streaming_integration_test.cc:500]: I removed many of these. I still have blank lines in a few places where I think it clarifies things for me, but not immediately around conditions, and I'll try to restrain my natural tendency for whitespace in the future.

File: source/extensions/filters/http/ext_proc/processor_state.cc (hunk 7 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/processor_state.cc]: So as you can see, the high limit is now the "bufferLimit" that comes back from the callbacks object, and the low limit is half that. I'll resolve this and you can re-open it if that doesn't match your understanding.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 5 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:213]: Added a longer comment with a link to the issue.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (gbrail) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: I added a method to convert the status to string -- only using it here but in the future maybe in other places.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by gbrail** [comment]
The original discussion about how to do this was covered in https://github.com/envoyproxy/envoy/issues/16760 and with this PR, that issue can be closed.

**Comment by machaval** [comment]
@gbrail do you have a working example of Streamed mode working I'm not able to make it work even on the most basic example... only on buffered mode

**Comment by gbrail** [comment]
Sure -- take a look at:

GoogleCloudPlatform/envoy-processor-examples (github.com)
<https://github.com/GoogleCloudPlatform/envoy-processor-examples>

In particular, there are two examples in the Go code of using streaming.
This one base64-encodes the HTTP response:

https://github.com/GoogleCloudPlatform/envoy-processor-examples/blob/f53b0d8698b76a0a586b60a42d3098d6b845dd14/go/cmd/processor/service.go#L611C9-L611C9

And this one calculates the hash of both the request and response by
reading the streaming body without modifying it:

https://github.com/GoogleCloudPlatform/envoy-processor-examples/blob/f53b0d8698b76a0a586b60a42d3098d6b845dd14/go/cmd/processor/service.go#L398
